File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
    if (source === destination) return true

    const graph = new Array(n).fill(null).map(() => [])

    for (const [u, v] of edges) {
        graph[u].push(v)
        graph[v].push(u)
    }

    const queue = [source]
    const visited = new Set()
    visited.add(source)

    while (queue.length > 0) {
        const current = queue.shift()
        if (current === destination) return true

        for (const neighbor of graph[current]) {
            if (!visited.has(neighbor)) {
                queue.push(neighbor)
                visited.add(neighbor)
            }
        }
    }

    return false
};

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        if source == destination:
            return True

        adj_list = [[] for _ in range(n)]

        for u, v in edges:
            adj_list[u].append(v)
            adj_list[v].append(u)

        visited = set()
        queue = deque([source])

        while queue:
            current = queue.popleft()
            if current == destination:
                return True
            
            for node in adj_list[current]:
                if node not in visited:
                    queue.append(node)
                    visited.add(node)


        return False